File: 0149-max-points-on-a-line/0149-max-points-on-a-line.py
class Solution:
    def maxPoints(self, pts: List[List[int]]) -> int:
        ans=1
        for i,(x, y) in enumerate(pts):
            d = defaultdict(int)
            for j in range(i+1, len(pts)):
                slope=(pts[j][1]-y)/(pts[j][0]-x) if pts[j][0]!=x else None
                d[slope]+=1
                ans=max(ans, d[slope]+1)
        return ans
        
        
# class Solution:
#     def maxPoints(self, pts: List[List[int]]) -> int:
        
#         def on_line(i, j, k):
#             if pts[j][0]-pts[i][0]==0:
#                 return pts[k][0]==pts[i][0]
#             if pts[j][1]-pts[i][1]==0:
#                 return pts[k][1]==pts[i][1]
#             return (pts[k][0]-pts[i][0])/(pts[j][0]-pts[i][0])==(pts[k][1]-pts[i][1])/(pts[j][1]-pts[i][1])
        
#         ans=1
#         for i,(x,y) in enumerate(pts):
#             done=set()
#             for j in range(i+1, len(pts)):
#                 if (pts[j][0], pts[j][1]) in done:
#                     continue
#                 satisf=2
#                 for k in range(j+1, len(pts)):
#                     if (pts[k][0], pts[k][1]) not in done and on_line(i, j, k):
#                         satisf+=1
#                         done.add((pts[k][0], pts[k][1]))
#                 ans=max(ans, satisf)
#         return ans
            
        

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.java
class Solution {
    public int maxPoints(int[][] pts) {
        int ans=1;
        for(int i=0; i<pts.length; ++i){
            HashMap<Double, Integer> map = new HashMap<>();
            for(int j=i+1; j<pts.length; ++j){
                Double slope = -0.0;
                if(pts[j][0] != pts[i][0]){
                    double up = (pts[j][1]-pts[i][1]);
                    double down = (pts[j][0]-pts[i][0]);
                    // slope = ((pts[j][1]-pts[i][1])/(pts[j][0]-pts[i][0]));
                    slope =  (up/down)==-0.0 ? 0.0 :(up/down);
                }
                if(map.containsKey(slope)) map.put(slope, map.get(slope)+1);
                else map.put(slope, 1);
                ans=Math.max(ans, map.get(slope)+1);
            }
        }
        return ans;
    }
}

// class Solution {
//     public int maxPoints(int[][] pts) {
//         int ans =1, n=pts.length;
//         for(int i=0; i<n; ++i){
//             HashSet<int[]> done = new HashSet<>();
//             for(int j=i+1; j<n; ++j){
//                 if(done.contains(new int[]{pts[j][0], pts[j][1]})) continue;
//                 int ptsOnLine=2;
//                 for(int k=j+1; k<n; k++){
//                     if(!done.contains(new int[]{pts[k][0], pts[k][1]}) && onLine(pts[i][0], pts[i][1], pts[j][0], pts[j][1], pts[k][0], pts[k][1])){
//                         ptsOnLine++;
//                         done.add(new int[]{pts[k][0], pts[k][1]});
//                     }
//                 }
//                 System.out.println(i+" "+j+" "+ptsOnLine);
//                 ans=Math.max(ans, ptsOnLine);
//             }
//         }
//         return ans;
//     }
//     private boolean onLine(int x1, int y1, int x2, int y2, int x, int y){
//         if(x2==x1) return x==x1;
//         if(y2==y1) return y==y1;
//         return (x-x1)/(x2-x1)==(y-y1)/(y2-y1);
//     }
// }